File: prettyd/src/ws.ts
import type { IncomingMessage } from 'node:http';
import type { Duplex } from 'node:stream';
import crypto from 'node:crypto';
import { WebSocketServer, type WebSocket } from 'ws';
import { getSession, writeInput, resize, type OutputEvent, type SessionHandle } from './sessions.js';
import type { ClaudeSessionEvent } from './sessionFileWatcher.js';
import { PROTOCOL_VERSION, type ClientMsg, type ServerMsg } from './types.js';
import { config, getAuthToken, isAllowedOrigin, isAuthOpen } from './config.js';

// Cap inbound frames at 256 KB — plenty for any realistic input (keystrokes,
// JSON control messages, paste). A client that exceeds it is either broken or
// malicious; ws closes the socket with code 1009 automatically.
const wss = new WebSocketServer({ noServer: true, maxPayload: 256 * 1024 });

// Locally extend ClientMsg to include ping. The types stream adds this to
// types.ts; the local union keeps ws.ts compiling in the interim.
type AnyClientMsg = ClientMsg | { type: 'ping' };

// 1 MB backpressure threshold. When the WS send buffer exceeds this the
// replay loop yields until the socket drains, preventing the daemon from
// blasting a multi-MB replay (e.g. the 156 MB output-history case) into
// the TCP send buffer all at once and stalling the Node event loop.
const MAX_BUFFERED = 1 * 1024 * 1024;

// Pause until ws.bufferedAmount drops below MAX_BUFFERED, with a 5-second
// safety timeout so a hung client can't block the loop indefinitely.
async function drainIfNeeded(ws: WebSocket): Promise<void> {
  if (ws.readyState !== ws.OPEN || ws.bufferedAmount <= MAX_BUFFERED) return;
  await new Promise<void>((resolve) => {
    const onDrain = (): void => { clearTimeout(timer); resolve(); };
    // If drain never fires (client died without closing), give up after 5 s.
    const timer = setTimeout(() => { ws.off('drain', onDrain); resolve(); }, 5_000);
    ws.once('drain', onDrain);
  });
}

function send(ws: WebSocket, msg: ServerMsg): void {
  if (ws.readyState === ws.OPEN) ws.send(JSON.stringify(msg));
}

// Attach one session's stream to a socket: replay (raw output from
// lastSeq, claude events from claudeEventsSince) then live events. Both
// connection modes go through here; mux mode tags every message with the
// sessionId so the client can route. Returns a detach() that removes the
// listeners, or null if the session already exited (terminal state was
// sent; nothing live to subscribe to).
//
// Made async so the replay loops can yield to the event loop via
// drainIfNeeded() when the WS send buffer is full — prevents the daemon
// from synchronously blasting a huge replay into one giant send-buffer
// burst (the "156 MB backlog" that caused 5-second typing lag).
async function attachSessionStream(
  ws: WebSocket,
  session: SessionHandle,
  opts: {
    lastSeq: number;
    claudeEventsSince: number;
    tagSessionId?: string;
    onExited?: () => void;
    // When false, raw PTY bytes are suppressed for this attach — no
    // replay AND no live output frames. Structured claudeEvents, exit,
    // and hello still flow. See the attach message in types.ts.
    includeOutput?: boolean;
    // When false, the on-attach replay of Claude JSONL *history* is
    // suppressed. Hidden sessions
    // attach this way so page load doesn't replay every session's
    // conversation at once. See the attach message in types.ts.
    includeClaudeReplay?: boolean;
    // When false, live structured claudeEvent frames are suppressed too.
    // Hidden sessions backfill with HTTP tail pages on activation, so
    // sending frames they immediately drop is pure background cost.
    includeClaudeLive?: boolean;
  }
): Promise<(() => void) | null> {
  const tag = opts.tagSessionId !== undefined ? { sessionId: opts.tagSessionId } : {};
  const includeOutput = opts.includeOutput !== false;
  const includeClaudeReplay = opts.includeClaudeReplay !== false;
  const includeClaudeLive = opts.includeClaudeLive !== false;
  const replay = session.log.since(opts.lastSeq);

  // Claude-event replay window. Indices are ABSOLUTE (claudeEventBase +
  // array index) so they survive the 5000-cap front-trim; map the
  // client's absolute resume point into a local offset. Fresh attaches
  // (since=0) cap the initial replay to the tail — long sessions hold
  // thousands of events and RemoteView renders ~50 anyway; older history
  // stays reachable via HTTP /events.
  const INITIAL_REPLAY_CAP = 300;
  const base = session.claudeEventBase;
  const len = session.claudeEventLog.length;
  const claudeEventsCount = base + len;
  // When history replay is suppressed, start the window at the current end
  // (localStart === len) so nothing historical is sent and the client's
  // counter is reported as already caught up. If live events are enabled,
  // they take it forward from there.
  const localStart = !includeClaudeReplay
    ? len
    : opts.claudeEventsSince > 0
      ? Math.max(0, Math.min(opts.claudeEventsSince - base, len))
      : Math.max(0, len - INITIAL_REPLAY_CAP);
  const claudeReplayStart = base + localStart;

  send(ws, {
    type: 'hello',
    protocol: PROTOCOL_VERSION,
    session: session.info,
    currentSeq: replay.current,
    resumedFromSeq: opts.lastSeq > 0 ? opts.lastSeq : null,
    claudeEventsCount,
    claudeReplayStart,
    ...tag
  });

  if (includeOutput) {
    if (replay.gap) {
      send(ws, { type: 'gap', oldestAvailableSeq: replay.oldest, currentSeq: replay.current, ...tag });
    }
    for (const ev of replay.events) {
      send(ws, { type: 'output', seq: ev.seq, data: ev.data, ...tag });
      // Yield when the send buffer fills up so we don't blast a
      // multi-MB output history into a single synchronous write burst.
      await drainIfNeeded(ws);
    }
  }
  for (let i = localStart; i < len; i++) {
    send(ws, { type: 'claudeEvent', event: session.claudeEventLog[i]!, ...tag });
    await drainIfNeeded(ws);
  }

  // Already exited: deliver the terminal state and skip live wiring.
  if (session.exited) {
    send(ws, {
      type: 'exit',
      code: session.exitCode,
      signal: session.exitSignal,
      seq: session.exitSeq ?? replay.current,
      ...tag
    });
    opts.onExited?.();
    return null;
  }

  const onOutput = (ev: OutputEvent): void => {
    send(ws, { type: 'output', seq: ev.seq, data: ev.data, ...tag });
  };
  const onExit = (info: { code: number | null; signal: string | null; seq: number }): void => {
    send(ws, { type: 'exit', code: info.code, signal: info.signal, seq: info.seq, ...tag });
    opts.onExited?.();
  };
  const onClaudeEvent = (ev: ClaudeSessionEvent): void => {
    send(ws, { type: 'claudeEvent', event: ev, ...tag });
  };
  // Contract #4: when the runner process disconnects while the session is
  // still alive (e.g. launchd killed the runner), emit a synthetic exit
  // frame so the browser's session view unfreezes rather than spinning
  // forever. code=null + reason='runner-lost' distinguishes this from a
  // real PTY exit; the client's existing exit handling path fires for both.
  const onRunnerLost = (): void => {
    if (ws.readyState === ws.OPEN) {
      ws.send(JSON.stringify({
        type: 'exit',
        code: null,
        signal: null,
        seq: replay.current,
        reason: 'runner-lost',
        ...tag
      }));
    }
  };
  if (includeOutput) session.emitter.on('output', onOutput);
  session.emitter.on('exit', onExit);
  if (includeClaudeLive) session.emitter.on('claudeEvent', onClaudeEvent);
  session.emitter.on('runner-lost', onRunnerLost);

  return () => {
    if (includeOutput) session.emitter.off('output', onOutput);
    session.emitter.off('exit', onExit);
    if (includeClaudeLive) session.emitter.off('claudeEvent', onClaudeEvent);
    session.emitter.off('runner-lost', onRunnerLost);
  };
}

// Clamp client-requested PTY dimensions. Last-resize-wins across clients:
// the most recently active viewport dictates the PTY size; the TUI
// redraws at the new size automatically.
function clampedResize(sessionId: string, cols: number, rows: number): void {
  resize(sessionId, Math.max(40, Math.min(500, cols | 0)), Math.max(10, Math.min(200, rows | 0)));
}

// Multiplexed mode (`/ws?mux=1`): one socket, many sessions (tmux-style).
// The client attaches/detaches sessions with tagged frames; every
// server→client message carries sessionId. N sessions cost ONE connection
// — the per-session-socket shape fell over once orchestrators pushed the
// session count past ~50 (dozens of parallel sockets, reconnect herds).
function handleMuxConnection(ws: WebSocket): void {
  const attached = new Map<string, () => void>(); // sessionId → detach
  // Guards against duplicate attach frames arriving while an async replay
  // is in progress (the await in attachSessionStream yields to the event
  // loop, so a second 'attach' for the same session can race through).
  const pendingAttach = new Set<string>();

  const detach = (sessionId: string): void => {
    const d = attached.get(sessionId);
    if (d) d();
    attached.delete(sessionId);
  };

  ws.on('message', (raw) => {
    // Frame-size enforcement is handled by maxPayload on the WebSocketServer
    // (code 1009 on overflow). This handler only fires for cleared frames.
    let parsed: AnyClientMsg | null = null;
    try {
      parsed = JSON.parse(raw.toString()) as AnyClientMsg;
    } catch {
      return; // mux mode is JSON-only; raw bytes have no session routing
    }
    // App-level keepalive ping (contract ws.d): JSON {type:'ping'} →
    // {type:'pong'}. Distinct from WebSocket's binary ping/pong opcodes;
    // this layer is what browser JS can actually send and receive.
    if (parsed.type === 'ping') {
      if (ws.readyState === ws.OPEN) ws.send(JSON.stringify({ type: 'pong' }));
      return;
    }
    if (parsed.type === 'attach') {
      const attachMsg = parsed; // capture narrowed type for the async closure
      const id = attachMsg.sessionId;
      if (!id || attached.has(id) || pendingAttach.has(id)) return;
      const session = getSession(id);
      if (!session) {
        send(ws, { type: 'error', message: `unknown session ${id}`, sessionId: id });
        return;
      }
      pendingAttach.add(id);
      void (async (): Promise<void> => {
        const det = await attachSessionStream(ws, session, {
          lastSeq: Math.max(0, (attachMsg.lastSeq ?? 0) | 0),
          claudeEventsSince: Math.max(0, (attachMsg.claudeEventsSince ?? 0) | 0),
          tagSessionId: id,
          includeOutput: attachMsg.outputReplay !== false,
          includeClaudeReplay: attachMsg.claudeReplay !== false,
          includeClaudeLive: attachMsg.claudeLive !== false,
          // On PTY exit only this session detaches — the socket stays up
          // for every other attached session.
          onExited: () => detach(id)
        });
        pendingAttach.delete(id);
        if (det) attached.set(id, det);
      })();
      return;
    }
    if (parsed.type === 'detach') {
      if (parsed.sessionId) detach(parsed.sessionId);
      return;
    }
    if (parsed.type === 'input') {
      if (parsed.sessionId) writeInput(parsed.sessionId, parsed.data);
      return;
    }
    if (parsed.type === 'resize') {
      if (parsed.sessionId) clampedResize(parsed.sessionId, parsed.cols, parsed.rows);
      return;
    }
  });

  ws.on('close', () => {
    for (const d of attached.values()) d();
    attached.clear();
  });
}

// Single-session mode (`/ws?sessionId=…`): the original protocol. Still
// used by the CLI (`pretty attach`, `pretty tail -f`) and any client
// that hasn't moved to mux.
//
// Message and close handlers are registered BEFORE awaiting the async
// replay so input arriving during a slow drain (large session history)
// is not silently discarded.
async function handleSingleConnection(ws: WebSocket, url: URL, sessionId: string): Promise<void> {
  const session = getSession(sessionId);
  if (!session) {
    send(ws, { type: 'error', message: `unknown session ${sessionId}` });
    ws.close(1008, 'unknown session');
    return;
  }

  const lastSeqParam = url.searchParams.get('lastSeq');
  const lastSeq = lastSeqParam ? Math.max(0, Number(lastSeqParam) | 0) : 0;
  const claudeSinceParam = url.searchParams.get('claudeEventsSince');
  const claudeEventsSince = claudeSinceParam ? Math.max(0, Number(claudeSinceParam) | 0) : 0;

  // detachFn is set after the async replay resolves; the close handler
  // holds a reference so it can clean up regardless of timing.
  let detachFn: (() => void) | null = null;

  ws.on('message', (raw, isBinary) => {
    if (isBinary) {
      writeInput(sessionId, (raw as Buffer).toString('utf8'));
      return;
    }
    const text = raw.toString();
    let parsed: AnyClientMsg | null = null;
    try {
      parsed = JSON.parse(text) as AnyClientMsg;
    } catch {
      writeInput(sessionId, text);
      return;
    }
    // App-level keepalive — see handleMuxConnection for rationale.
    if (parsed.type === 'ping') {
      if (ws.readyState === ws.OPEN) ws.send(JSON.stringify({ type: 'pong' }));
      return;
    }
    if (parsed.type === 'input') {
      writeInput(sessionId, parsed.data);
    } else if (parsed.type === 'resize') {
      clampedResize(sessionId, parsed.cols, parsed.rows);
    }
  });

  ws.on('close', () => {
    detachFn?.();
  });

  detachFn = await attachSessionStream(ws, session, {
    lastSeq,
    claudeEventsSince,
    onExited: () => ws.close(1000, 'pty exited')
  });
  // null means the session was already exited when we attached — the exit
  // frame and onExited callback were handled inside attachSessionStream.
}

wss.on('connection', (ws, req) => {
  const url = new URL(req.url ?? '/', 'http://localhost');
  if (url.searchParams.get('mux') === '1') {
    handleMuxConnection(ws);
    return;
  }
  const sessionId = url.searchParams.get('sessionId');
  if (!sessionId) {
    send(ws, { type: 'error', message: 'missing sessionId' });
    ws.close(1008, 'missing sessionId');
    return;
  }
  void handleSingleConnection(ws, url, sessionId);
});

export function handleUpgrade(req: IncomingMessage, socket: Duplex, head: Buffer): void {
  const url = new URL(req.url ?? '/', 'http://localhost');
  if (url.pathname !== '/ws') {
    socket.destroy();
    return;
  }

  // Reject cross-origin browser requests BEFORE completing the WebSocket
  // handshake — CSRF mitigation. Non-browser clients (CLI, curl) send no
  // Origin and are always allowed.
  if (!isAllowedOrigin(req.headers.origin, config.host)) {
    socket.write('HTTP/1.1 403 Forbidden\r\nContent-Length: 0\r\n\r\n');
    socket.destroy();
    return;
  }

  // Validate the auth token from the ?token= query param. Browsers cannot
  // set custom WebSocket headers, so the token travels in the URL per
  // contract #5. Use timingSafeEqual to avoid timing-oracle attacks.
  const qToken = url.searchParams.get('token') ?? '';
  const authToken = getAuthToken();
  const authBuf = Buffer.from(authToken, 'utf8');
  const qBuf = Buffer.from(qToken, 'utf8');
  // Length check short-circuits before timingSafeEqual (equal-length
  // requirement) but still leaks no timing info about the token value.
  const tokenOk = qBuf.length > 0 &&
    qBuf.length === authBuf.length &&
    crypto.timingSafeEqual(qBuf, authBuf);
  if (!tokenOk && !isAuthOpen()) {
    // Respond with HTTP 401 before destroying the socket. Using close
    // code 4401 post-upgrade would require completing the WS handshake
    // first, which we avoid so the socket is never treated as live.
    socket.write('HTTP/1.1 401 Unauthorized\r\nContent-Length: 0\r\n\r\n');
    socket.destroy();
    return;
  }

  wss.handleUpgrade(req, socket, head, (ws) => {
    wss.emit('connection', ws, req);
  });
}

File: prettygo/internal/api/ws.go
package api

import (
	"context"
	"encoding/json"
	"math"
	"net/http"
	"strconv"
	"sync"
	"time"

	"github.com/coder/websocket"
	"github.com/uzihaq/pretty-pty/prettygo/internal/proto"
	"github.com/uzihaq/pretty-pty/prettygo/internal/state"
)

const (
	websocketReadLimit = 256 * 1024
	clientProtocol     = 2
)

type wsPeer struct {
	connection *websocket.Conn
	writes     sync.Mutex
}

func (p *wsPeer) send(ctx context.Context, value any) error {
	encoded, err := json.Marshal(value)
	if err != nil {
		return err
	}
	p.writes.Lock()
	defer p.writes.Unlock()
	writeCtx, cancel := context.WithTimeout(ctx, 5*time.Second)
	defer cancel()
	return p.connection.Write(writeCtx, websocket.MessageText, encoded)
}

func (s *Server) serveWebSocket(response http.ResponseWriter, request *http.Request) {
	connection, err := websocket.Accept(response, request, &websocket.AcceptOptions{
		InsecureSkipVerify: true, // Origin was checked against config.ts parity above.
		CompressionMode:    websocket.CompressionDisabled,
	})
	if err != nil {
		return
	}
	connection.SetReadLimit(websocketReadLimit)
	peer := &wsPeer{connection: connection}
	if request.URL.Query().Get("mux") == "1" {
		s.handleMux(request.Context(), peer)
		return
	}
	s.handleSingle(request.Context(), peer, request)
}

func (s *Server) handleSingle(parent context.Context, peer *wsPeer, request *http.Request) {
	ctx, cancel := context.WithCancel(parent)
	defer cancel()
	defer peer.connection.CloseNow()

	id := request.URL.Query().Get("sessionId")
	if id == "" {
		_ = peer.send(ctx, map[string]any{"type": "error", "message": "missing sessionId"})
		_ = peer.connection.Close(websocket.StatusPolicyViolation, "missing sessionId")
		return
	}
	session, ok := s.registry.Get(id)
	if !ok {
		_ = peer.send(ctx, map[string]any{"type": "error", "message": "unknown session " + id})
		_ = peer.connection.Close(websocket.StatusPolicyViolation, "unknown session")
		return
	}
	lastSeq := nonnegativeUint(request.URL.Query().Get("lastSeq"))
	claudeSince := int64(nonnegativeUint(request.URL.Query().Get("claudeEventsSince")))
	attachment := session.Attach(state.AttachOptions{
		LastSeq: lastSeq, ClaudeEventsSince: claudeSince,
		IncludeClaudeReplay: true, InitialReplayCap: 300,
	})
	defer attachment.Cancel()
	if err := sendInitial(ctx, peer, session, attachment, "", lastSeq, true); err != nil {
		return
	}
	if exited, terminal := session.TerminalState(); exited {
		_ = peer.send(ctx, exitMessage(terminal, ""))
		_ = peer.connection.Close(websocket.StatusNormalClosure, "pty exited")
		return
	}

	go streamAttachment(ctx, peer, attachment, streamOptions{
		includeOutput: true, includeClaudeLive: true,
		onExit: func() {
			_ = peer.connection.Close(websocket.StatusNormalClosure, "pty exited")
			cancel()
		},
	})
	for {
		messageType, payload, err := peer.connection.Read(ctx)
		if err != nil {
			return
		}
		if messageType == websocket.MessageBinary {
			session.Input(ctx, string(payload))
			continue
		}
		var message clientMessage
		if err := json.Unmarshal(payload, &message); err != nil {
			session.Input(ctx, string(payload))
			continue
		}
		switch message.Type {
		case "ping":
			_ = peer.send(ctx, map[string]any{"type": "pong"})
		case "input":
			session.Input(ctx, message.Data)
		case "resize":
			session.Resize(ctx, clampDimension(message.Cols, 40, 500), clampDimension(message.Rows, 10, 200))
		}
	}
}

type muxAttachment struct {
	cancel func()
}

func (s *Server) handleMux(parent context.Context, peer *wsPeer) {
	ctx, cancel := context.WithCancel(parent)
	defer cancel()
	defer peer.connection.CloseNow()
	attached := make(map[string]muxAttachment)
	var attachedMu sync.Mutex
	detach := func(id string) {
		attachedMu.Lock()
		entry, ok := attached[id]
		if ok {
			delete(attached, id)
		}
		attachedMu.Unlock()
		if ok {
			entry.cancel()
		}
	}
	defer func() {
		attachedMu.Lock()
		entries := make([]muxAttachment, 0, len(attached))
		for _, entry := range attached {
			entries = append(entries, entry)
		}
		attached = make(map[string]muxAttachment)
		attachedMu.Unlock()
		for _, entry := range entries {
			entry.cancel()
		}
	}()

	for {
		messageType, payload, err := peer.connection.Read(ctx)
		if err != nil {
			return
		}
		if messageType != websocket.MessageText {
			continue
		}
		var message clientMessage
		if err := json.Unmarshal(payload, &message); err != nil {
			continue
		}
		switch message.Type {
		case "ping":
			_ = peer.send(ctx, map[string]any{"type": "pong"})
		case "attach":
			if message.SessionID == "" {
				continue
			}
			attachedMu.Lock()
			_, exists := attached[message.SessionID]
			attachedMu.Unlock()
			if exists {
				continue
			}
			session, ok := s.registry.Get(message.SessionID)
			if !ok {
				_ = peer.send(ctx, map[string]any{
					"type": "error", "message": "unknown session " + message.SessionID,
					"sessionId": message.SessionID,
				})
				continue
			}
			includeOutput := message.OutputReplay == nil || *message.OutputReplay
			includeClaudeReplay := message.ClaudeReplay == nil || *message.ClaudeReplay
			includeClaudeLive := message.ClaudeLive == nil || *message.ClaudeLive
			attachment := session.Attach(state.AttachOptions{
				LastSeq: message.LastSeq, ClaudeEventsSince: message.ClaudeEventsSince,
				IncludeClaudeReplay: includeClaudeReplay, InitialReplayCap: 300,
			})
			attachedMu.Lock()
			attached[message.SessionID] = muxAttachment{cancel: attachment.Cancel}
			attachedMu.Unlock()
			if err := sendInitial(ctx, peer, session, attachment, message.SessionID, message.LastSeq, includeOutput); err != nil {
				detach(message.SessionID)
				continue
			}
			if exited, terminal := session.TerminalState(); exited {
				_ = peer.send(ctx, exitMessage(terminal, message.SessionID))
				detach(message.SessionID)
				continue
			}
			id := message.SessionID
			go streamAttachment(ctx, peer, attachment, streamOptions{
				sessionID: id, includeOutput: includeOutput, includeClaudeLive: includeClaudeLive,
				onExit: func() { detach(id) },
			})
		case "detach":
			if message.SessionID != "" {
				detach(message.SessionID)
			}
		case "snapshot":
			s.handleMuxSnapshot(ctx, peer, message)
		case "events":
			s.handleMuxEvents(ctx, peer, message)
		case "input":
			session, ok := s.registry.Get(message.SessionID)
			written := ok && session.Input(ctx, message.Data)
			if message.RequestID != "" {
				_ = peer.send(ctx, map[string]any{
					"type": "inputAck", "requestId": message.RequestID,
					"sessionId": message.SessionID, "ok": written,
				})
			}
		case "resize":
			if session, ok := s.registry.Get(message.SessionID); ok {
				session.Resize(ctx, clampDimension(message.Cols, 40, 500), clampDimension(message.Rows, 10, 200))
			}
		}
	}
}

type clientMessage struct {
	Type              string `json:"type"`
	Data              string `json:"data"`
	SessionID         string `json:"sessionId"`
	RequestID         string `json:"requestId"`
	Cols              int    `json:"cols"`
	Rows              int    `json:"rows"`
	LastSeq           uint32 `json:"lastSeq"`
	ClaudeEventsSince int64  `json:"claudeEventsSince"`
	Since             *int64 `json:"since"`
	Tail              *int64 `json:"tail"`
	OutputReplay      *bool  `json:"outputReplay"`
	ClaudeReplay      *bool  `json:"claudeReplay"`
	ClaudeLive        *bool  `json:"claudeLive"`
}

func sendInitial(
	ctx context.Context,
	peer *wsPeer,
	session *state.Session,
	attachment state.Attachment,
	sessionID string,
	lastSeq uint32,
	includeOutput bool,
) error {
	resumed := any(nil)
	if lastSeq > 0 {
		resumed = lastSeq
	}
	hello := map[string]any{
		"type": "hello", "protocol": clientProtocol, "session": session.Info(),
		"currentSeq": attachment.Replay.Current, "resumedFromSeq": resumed,
		"claudeEventsCount": attachment.ClaudeEventsCount,
		"claudeReplayStart": attachment.ClaudeReplayStart,
	}
	if sessionID != "" {
		hello["sessionId"] = sessionID
	}
	if err := peer.send(ctx, hello); err != nil {
		return err
	}
	if includeOutput {
		if attachment.Replay.Gap {
			message := map[string]any{
				"type": "gap", "oldestAvailableSeq": attachment.Replay.Oldest,
				"currentSeq": attachment.Replay.Current,
			}
			addSessionID(message, sessionID)
			if err := peer.send(ctx, message); err != nil {
				return err
			}
		}
		for _, event := range attachment.Replay.Events {
			message := map[string]any{"type": "output", "seq": event.Seq, "data": event.Data}
			addSessionID(message, sessionID)
			if err := peer.send(ctx, message); err != nil {
				return err
			}
		}
	}
	for _, event := range attachment.ClaudeEvents {
		message := map[string]any{"type": "claudeEvent", "event": json.RawMessage(event)}
		addSessionID(message, sessionID)
		if err := peer.send(ctx, message); err != nil {
			return err
		}
	}
	return nil
}

type streamOptions struct {
	sessionID         string
	includeOutput     bool
	includeClaudeLive bool
	onExit            func()
}

func streamAttachment(ctx context.Context, peer *wsPeer, attachment state.Attachment, options streamOptions) {
	for event := range attachment.Events {
		var message map[string]any
		switch event.Kind {
		case proto.EventOutput:
			if !options.includeOutput || event.Output.Seq <= attachment.Replay.Current {
				continue
			}
			message = map[string]any{"type": "output", "seq": event.Output.Seq, "data": event.Output.Data}
		case proto.EventClaude:
			if !options.includeClaudeLive || event.ClaudeIndex < attachment.ClaudeEventsCount {
				continue
			}
			message = map[string]any{"type": "claudeEvent", "event": json.RawMessage(event.ClaudeEvent)}
		case proto.EventExit, proto.EventRunnerLost:
			message = exitMessage(event.Exit, options.sessionID)
			if err := peer.send(ctx, message); err == nil && options.onExit != nil {
				options.onExit()
			}
			return
		default:
			continue
		}
		addSessionID(message, options.sessionID)
		if err := peer.send(ctx, message); err != nil {
			return
		}
	}
}

func (s *Server) handleMuxSnapshot(ctx context.Context, peer *wsPeer, message clientMessage) {
	if message.RequestID == "" || message.SessionID == "" {
		return
	}
	session, ok := s.registry.Get(message.SessionID)
	if !ok {
		sendRPCError(ctx, peer, message.RequestID, "unknown session "+message.SessionID, "not_found", message.SessionID)
		return
	}
	cols := message.Cols
	if cols < 0 {
		cols = 0
	}
	text, seq, err := session.Snapshot(ctx, cols)
	if err != nil {
		sendRPCError(ctx, peer, message.RequestID, err.Error(), "", message.SessionID)
		return
	}
	_ = peer.send(ctx, map[string]any{
		"type": "snapshot", "requestId": message.RequestID, "sessionId": message.SessionID,
		"text": text, "seq": seq,
	})
}

func (s *Server) handleMuxEvents(ctx context.Context, peer *wsPeer, message clientMessage) {
	if message.RequestID == "" || message.SessionID == "" {
		return
	}
	session, ok := s.registry.Get(message.SessionID)
	if !ok {
		sendRPCError(ctx, peer, message.RequestID, "unknown session "+message.SessionID, "not_found", message.SessionID)
		return
	}
	window := session.EventsWindow(message.Since, message.Tail, nil)
	_ = peer.send(ctx, map[string]any{
		"type": "events", "requestId": message.RequestID, "sessionId": message.SessionID,
		"events": window.Events, "nextIndex": window.NextIndex, "totalCount": window.TotalCount,
	})
}

func sendRPCError(ctx context.Context, peer *wsPeer, requestID, message, code, sessionID string) {
	response := map[string]any{"type": "rpcError", "requestId": requestID, "message": message}
	if code != "" {
		response["code"] = code
	}
	if sessionID != "" {
		response["sessionId"] = sessionID
	}
	_ = peer.send(ctx, response)
}

func exitMessage(exit proto.ExitEvent, sessionID string) map[string]any {
	message := map[string]any{
		"type": "exit", "code": exit.Code, "signal": exit.Signal, "seq": exit.Seq,
	}
	if exit.Reason != "" {
		message["reason"] = exit.Reason
	}
	addSessionID(message, sessionID)
	return message
}

func addSessionID(message map[string]any, sessionID string) {
	if sessionID != "" {
		message["sessionId"] = sessionID
	}
}

func nonnegativeUint(raw string) uint32 {
	if raw == "" {
		return 0
	}
	value, err := strconv.ParseFloat(raw, 64)
	if err != nil || math.IsNaN(value) || math.IsInf(value, 0) || value <= 0 {
		return 0
	}
	if value > float64(^uint32(0)) {
		return ^uint32(0)
	}
	return uint32(value)
}

func clampDimension(value, minimum, maximum int) int {
	if value < minimum {
		return minimum
	}
	if value > maximum {
		return maximum
	}
	return value
}
